### PR TITLE
Fix race for concurrent downloads of remote media. 

### LINKED ISSUE #6755: duplicate key value violates unique constraint "remote_media_cache_media_origin_media_id_key" when getting remote thumbnails
on matrix-org-hotfixes:

```
psycopg2.errors.UniqueViolation: duplicate key value violates unique constraint "remote_media_cache_media_origin_media_id_key"
DETAIL:  Key (media_origin, media_id)=(XXX.net, XXXXXXXXXXXXscjwZ) already exists.
```
This fires when attempting to get a thumbnail for some remote media:

`GET /_matrix/media/r0/thumbnail/XXX.net/XXXXXXXXXXXXscjwZ?width=30&height=30&method=crop`

There were multiple requests to this URL from different IPs near-simultaneously - presumably someone just posted this to a large channel.

The request to the remote server to download the media looks just fine, I assume there's a race somewhere. If this duplicate error is OK can we suppress the logging of it, as it's a ~200line stacktrace, but ideally the additional connections block and don't process the upload.

### PR DESCRIPTION
**Pull request by erikjohnston**
Fixes #6755

### CODE REVIEW

**Comment by erikjohnston** [comment]
FTR the idea here is to move the DB calls into `with store_into_file`, which will correctly delete the temporary file if an exception occurs. We then catch any exceptions and check if we magically now have the requested media downloaded after all.

File: synapse/rest/media/v1/media_repository.py [hunk 1]
```diff
@@ -324,14 +321,28 @@ async def _get_remote_media_impl(
...
+            media_info = await self.store.get_cached_remote_media(server_name, media_id)
+            if not media_info:
+                raise e
+
+        file_id = media_info["filesystem_id"]
+        file_info = FileInfo(server_name, file_id)
+
+        await self._generate_thumbnails(
+            server_name, media_id, file_id, media_info["media_type"]
+        )
```
Comment (clokep) [inline synapse/rest/media/v1/media_repository.py:340]: Shouldn't we only do this when `_download_remote_file` is successful?

If the media magically showed up will this be a waste of effort / conflict / raise an error that we're trying to re-generate thumbnails?

(It looks like this used to be in `download_remote_file`, why was this moved?)

File: synapse/rest/media/v1/media_repository.py [hunk 2]
```diff
@@ -401,22 +414,32 @@ async def _download_remote_file(
...
+            # Multiple remote media download requests can race (when using
+            # multiple media repos), so this may throw a violation constraint
+            # exception. If it does we'll delete the newly downloaded file from
+            # disk (as we're in the ctx manager).
+            #
+            # However: we've already called `finish()` so we may have also
+            # written to the storage providers. This is preferable to the
+            # alternative where we call `finish()` *after* this, where we could
+            # end up having an entry in the DB but fail to write the files to
+            # the storage providers.
```
Comment (clokep) [inline synapse/rest/media/v1/media_repository.py:430]: Meaning we might still end up with files that are useless?

File: synapse/rest/media/v1/media_repository.py (hunk 1 above)
Comment (erikjohnston) [reply synapse/rest/media/v1/media_repository.py:340]: I should add a comment, but broadly: 1) it's conceivable that the other download request dies before it generates thumbnails, but mainly b) we want toe be sure the thumbnails have finished being generated before responding to the client, otherwise they'll request thumbnails and get a 404 if they're not ready yet.

File: synapse/rest/media/v1/media_repository.py (hunk 2 above)
Comment (erikjohnston) [reply synapse/rest/media/v1/media_repository.py:430]: Yes, we should file an issue to fix that though (e.g adding a `delete_file` function to the storage providers and calling that).

File: synapse/rest/media/v1/media_repository.py (hunk 1 above)
Comment (erikjohnston) [reply synapse/rest/media/v1/media_repository.py:346]: We moved it out because we want it to get run even if `store_cached_remote_media` throws an duplicate key error

Comment (clokep) [reply synapse/rest/media/v1/media_repository.py:346]: Right. So if you have two that are racing and both work, you might get a dup key error for the media, then a dup key error for the thumbnail?

Comment (clokep) [reply synapse/rest/media/v1/media_repository.py:346]: We talked through this in #synapse-dev:matrix.org and pretty much we want to ensure the thumbnails exist before existing. So we try it and potentially capture the error in the same way if it already exists. Which is a naive implementation that wastes a bit of work.

File: synapse/rest/media/v1/media_storage.py [hunk 3]
```diff
@@ -112,14 +116,20 @@ def store_into_file(self, file_info: FileInfo):
...
         finished_called = [False]
 
-        async def finish():
-            for provider in self.storage_providers:
-                await provider.store_file(path, file_info)
-
-            finished_called[0] = True
-
         try:
             with open(fname, "wb") as f:
```
Comment (clokep) [inline synapse/rest/media/v1/media_storage.py:120]: Huh, this being a context manager here is a bit weird when we don't really write inside of it...

Comment (erikjohnston) [reply synapse/rest/media/v1/media_storage.py:120]: Well, we do in that we write during the `yield`, as the function is a context manager?

**Review by clokep — APPROVED** [review]
Looks good! Can you file the follow-up mentioned?

File: synapse/rest/media/v1/media_repository.py [hunk 4]
```diff
@@ -324,14 +321,34 @@ async def _get_remote_media_impl(
...
+            media_info = await self.store.get_cached_remote_media(server_name, media_id)
+            if not media_info:
+                raise e
+
+        file_id = media_info["filesystem_id"]
+        file_info = FileInfo(server_name, file_id)
+
+        # We generate thumbnails event if another process downloaded the media
+        # as a) it's conceivable that the other download request dies before it
+        # generates thumbnails, but mainly b) we want toe be sure the thumbnails
```
Comment (clokep) [inline synapse/rest/media/v1/media_repository.py:340]:
```suggestion
        # We generate thumbnails even if another process downloaded the media
        # as a) it's conceivable that the other download request dies before it
        # generates thumbnails, but mainly b) we want to be sure the thumbnails
```
